test/js/visualization/scatter-plot.test.js

```diff
@@ -1,43 +1,112 @@
+import ScatterPlot, { getPlotlyTraces } from 'components/visualization/ScatterPlot'
+import * as ScpApiMetrics from 'lib/scp-api-metrics'
+import * as MetricsApi from 'lib/metrics-api'
 
-const BASIC_PLOT_DATA = {
-  axes: {
-    aspects: null,
-    titles: {
-      magnitude: 'Expression',
-      x: 'X',
-      y: 'Y',
-      z: 'Z'
-    }
-  },
-  scatter: {
-    annotParams: {
-      name: 'Category',
-      type: 'group',
-      scope: 'cluster'
-    },
-    genes: ['foo']
-  },
-  isAnnotatedScatter: false,
-  isCorrelatedScatter: false,
-  scatterColor: '',
-  dataScatterColor: undefined,
-  pointSize: 3,
-  pointAlpha: 1,
-  is3d: false,
-  annotType: 'group',
-  annotName: 'biosample_id',
-  data: {
-    x: [1, 2, 3, 4, 5, 6, 7, 8],
-    y: [1, 2, 3, 4, 5, 6, 7, 8],
-    cells: ['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H'],
-    annotations: ['s1', 's1', 's1', 's1', 's2', 's2', 's1', 's2']
-  }
+import '@testing-library/jest-dom/extend-expect'
+
+import { BASIC_PLOT_DATA, MANY_LABELS_MOCKS } from './scatter-plot.test-data'
+
+const CACHE_PERF_PARAMS = {
+  legacyBackend: 0,
+  parse: 0,
+  url: 'cache'
 }
 
+const BASIC_DIMENSION_PROPS = {
+  isMultiRow: false,
+  isTwoColumn: false,
+  showRelatedGenesIdeogram: false,
+  showViewOptionsControls: true
+}
+
+beforeAll(() => {
+  global.$ = jquery
+})
+// Note: tests that mock global.fetch must be cleared after every test
+afterEach(() => {
+  // Restores all mocks back to their original value
+  jest.restoreAllMocks()
+})
+
+
+it('shows custom legend with default group scatter plot', async () => {
+  const apiFetch = jest.spyOn(ScpApi, 'fetchCluster')
+  // pass in a clone of the response since it may get modified by the cache operations
+  apiFetch.mockImplementation(params => {
+    const response = _cloneDeep(MANY_LABELS_MOCKS.CLUSTER_RESPONSE)
+    response.cluster = params.cluster
+    response.genes = params.genes
+    return Promise.resolve([response, CACHE_PERF_PARAMS])
+  })
+  const fakePlot = jest.spyOn(Plotly, 'react')
+  fakePlot.mockImplementation(() => {})
+  const fakeLogScatter = jest.spyOn(ScpApiMetrics, 'logScatterPlot')
+  fakeLogScatter.mockImplementation(() => {})
+
+  const fakeLog = jest.spyOn(MetricsApi, 'log')
+  fakeLog.mockImplementation(() => {})
+
+  const { container } = render((
+    <ScatterPlot studyAccession='SCP101'
+      {...{
+        cluster: 'cluster_many_long_odd_labels.tsv',
+        annotation: {
+          name: 'Category',
+          type: 'group',
+          scope: 'cluster'
+        },
+        subsample: 'all',
+        consensus: null,
+        genes: [],
+        dimensionProps: BASIC_DIMENSION_PROPS
+      }}/>
+  ))
+
+  // findByTestId would be more kosher, but the numerical leaf ("1") is
+  // assigned randomly in `ScatterPlot`.  This ID has been observed with the
+  // value "study-scatter-25", but relying on that seems brittle.
+  //
+  // Consider changing `ScatterPlot` to use a deterministic `data-testid`.
```

Comment:
I agree with this comment - having a function that would deterministically assign DOM ids would be more robust/testable.